triage/experiments/validate.py

```diff
@@ -0,0 +1,221 @@
+            column_should_be_intlike(dense_state_table, 'entity_id', self.db_engine)
+            column_should_be_stringlike(dense_state_table, 'state', self.db_engine)
+            column_should_be_timelike(dense_state_table, 'start_time', self.db_engine)
+            column_should_be_timelike(dense_state_table, 'end_time', self.db_engine)
+            if 'state_filters' not in state_config or len(state_config['state_filters']) < 1:
+                raise ValueError('If a table_name is given in state_config, at least one state filter must be present')
+
+
+class FeatureGroupDefinitionValidator(Validator):
+    def run(self, feature_group_definition, feature_aggregation_config):
+        if not isinstance(feature_group_definition, dict):
+            raise ValueError('Feature Group Definition must be a dictionary')
+
+        available_subsetters = architect.feature_group_creator.FeatureGroupCreator.subsetters
+        for subsetter_name, value in feature_group_definition.items():
+            if subsetter_name not in available_subsetters:
+                raise ValueError('''Unknown feature_group_definition key {} received.
+                Available keys are {}'''.format(subsetter_name, available_subsetters))
+            if not hasattr(value, '__iter__') or isinstance(value, (str, bytes)):
+                raise ValueError('Each value in FeatureGroupCreator must be iterable and not a string')
+
+        if 'prefix' in feature_group_definition:
+            available_prefixes = [
+                aggregation['prefix']
+                for aggregation in feature_aggregation_config
+            ]
+            for prefix in feature_group_definition['prefix']:
+                if prefix not in available_prefixes:
+                    raise ValueError('''Aggregation prefix of '{}' was given as a
+                    feature group, but no such prefix exists in the available
+                    feature aggregations. The available feature aggregations are {}
+                    '''.format(prefix, available_prefixes))
+
+class FeatureGroupStrategyValidator(Validator):
+    def run(self, feature_group_strategies):
+        if not isinstance(feature_group_strategies, list):
+            raise ValueError('feature_group_strategies key must be a list')
+        available_strategies = architect.feature_group_mixer.FeatureGroupMixer.strategy_lookup
+        for strategy in feature_group_strategies:
+            if strategy not in available_strategies:
+                raise ValueError('''Unknown feature_group_strategies key {} received.
+                Available keys are {}'''.format(strategy, available_strategies))
+
+
+class UserMetadataValidator(Validator):
+    def run(self, user_metadata):
+        if not isinstance(user_metadata, dict):
+            raise ValueError('user_metadata key must be a dict')
+
+
+class ModelGroupKeysValidator(Validator):
+    def run(self, model_group_keys, user_metadata):
```

Comment:
I think there might be more we can do here around potential logical fallacies or places where the user might get unexpected results, at least with warnings. For instance, I can't think of a good reason not to have `train_duration` as part of the model group keys, while at the same time, I can't think of a case where having `matrix_end_time` in the keys would make sense.

Reply:
This should be good, but this seems it might be better to actually bar them from doing certain things, or auto-adding certain things. If there isn't a good reason not to have certain keys, should we just add them to the defaults?

Reply:
yeah, adding to defaults and disallowing certain things might make more sense than simply checking for those things on validation.